File: Week Assignment 4 using C/7_check_circular_ll.c
// 7. Check if linked list is circular (returns 1 if circular else 0)
#include <stdio.h>
#include <stdlib.h>

struct Node { int data; struct Node* next; };

int isCircular(struct Node* head) {
    if (!head) return 1; // empty list is considered circular per many definitions
    struct Node* cur = head->next;
    while (cur && cur != head) cur = cur->next;
    return cur == head;
}

int main() {
    struct Node* a = (struct Node*)malloc(sizeof(struct Node)); a->data=1; a->next=NULL;
    struct Node* b = (struct Node*)malloc(sizeof(struct Node)); b->data=2; b->next=NULL;
    a->next = b; b->next = a; // circular
    printf("%d\n", isCircular(a)); // 1
    b->next = NULL;
    printf("%d\n", isCircular(a)); // 0
    return 0;
}

File: Week Assignement 4 using Python/7_check_circular_ll.py
# 7. Check if linked list is circular
class Node:
    def __init__(self,data):
        self.data=data; self.next=None

def is_circular(head):
    if not head: return True
    cur = head.next
    while cur and cur != head: cur = cur.next
    return cur == head

if __name__ == "__main__":
    a = Node(1); b = Node(2); a.next = b; b.next = a
    print(is_circular(a))
    b.next = None
    print(is_circular(a))
